lib/RestructureCFG/CycleEquivalenceAnalysis.cpp

```diff
@@ -0,0 +1,659 @@
+//
+// Copyright rev.ng Labs Srl. See LICENSE.md for details.
+//
+
+#include <iterator>
+#include <limits>
+#include <map>
+
+#include "llvm/ADT/DepthFirstIterator.h"
+#include "llvm/ADT/PostOrderIterator.h"
+
+#include "revng/ADT/GenericGraph.h"
+#include "revng/Support/Debug.h"
+#include "revng/Support/IRHelpers.h"
+
+#include "revng-c/RestructureCFG/CycleEquivalenceAnalysis.h"
+
+using namespace llvm;
+using namespace llvm::cl;
+
+// Debug logger.
+Logger<> CycleEquivalenceAnalysisLogger("cycle-equivalence");
+
+template<class GraphT, class GT>
+void CycleEquivalenceAnalysis<GraphT, GT>::insertEdge(BlockEdgeDescriptor E,
+                                                      EdgeLabel *Label,
+                                                      bool IsInverted,
```

Comment:
It seems a little bit awkward to compute `IsInverted` very early (in `getTreeEdges`, `getBackwardEdges` ...) and make it propagate until here.

I think that `getTreeEdges` and friends could avoid to compute that and just return edge iterators that point to the `Undirected` edges.

Then, nobody else uses the `IsInverted` field, and the edge iterators end up being passed in here instead of `BlockEdgeDescriptor` and `EdgeLabel` (both can be recovered from the edge iterator IIUC).

So the computation of `IsInverted` can be centralized here, which is the only place where we care about I think. And perhaps you can even drop the `BlockEdgeDescriptor` alltogether.

Reply:
Mmmh, I have some doubt on how to implement this though:

You suggested returning the edge iterators from `getTreeEdges`, and I also think that it would be okay in all the call sites, except in the `insertEdge` method, where I actually need the `IsInverted` property.

Indeed, how can I compute the `IsInverted` here, just from the edge iterators?
Assume that I'm in a invocation of the `insertEdge`, I get as a parameter the edge iterator, and I want to compute if the passed edge iterator refers to a `Inverted` edge. Doing such thing would mean checking if the current edge, when we "see" the graph as `Undirected" is part of the "predecessor" half, meaning it has surpassed the "successor" half (remember that we `concat(successors + predecessors)`. I cannot compare the edge iterator passed by parameter with the iterator of the last edge which is not inverted not applying the `Undirected` trait.

But even admitting that I pass both the edge iterator of the edge we want to insert, and the iterator pointing to the last edge which is not inverted (and I would need to return both such iterators from the `getTreeEdges`, which is not super elegant), I cannot compare with `<` iterators that are not `random_access_iterator`s.

Therefore, the only way to compute the `IsInverted`, is to start from zero again and iterate over all the successors, so we can keep track of the `IsInverted` property using the `Index` of the successors as it is done now in the `getTreeEdges` method.

What am I missing with respect to what you suggested?

Reply:
Yeah your line of reasoning makes sense. Thanks for walking me through it.